snappy_pipeline/workflows/adapter_trimming/__init__.py

```diff
@@ -0,0 +1,334 @@
+
+    @classmethod
+    @dictify
+    def get_input_files(cls, action):
+        """Return input files"""
+        assert action == "run"
+        yield "done", "work/input_links/{library_name}/.done"
+
+    @dictify
+    def get_output_files(self, action):
+        """Return output files"""
+        assert action == "run"
+        return (
+            ("out_done", self.base_path_out.format(trimmer=self.name) + "/out/.done"),
+            ("report_done", self.base_path_out.format(trimmer=self.name) + "/report/.done"),
+        )
+
+    @dictify
+    def _get_log_file(self, action):
+        """Return dict of log files."""
+        _ = action
+        prefix = "work/{trimmer}.{{library_name}}/log/{trimmer}.{{library_name}}".format(
+            trimmer=self.__class__.name
+        )
+        key_ext = (
+            ("log", ".log"),
+            ("conda_info", ".conda_info.txt"),
+            ("conda_list", ".conda_list.txt"),
+        )
+        yield "done", "work/{trimmer}.{{library_name}}/log/.done".format(
+            trimmer=self.__class__.name
+        )
+        for key, ext in key_ext:
+            yield key, prefix + ext
+            yield key + "_md5", prefix + ext + ".md5"
+
+    def get_args(self, action):
+        """Return function that maps wildcards to dict for input files"""
+
+        def args_function(wildcards):
+            result = {
+                "input": {
+                    "reads_left": list(
+                        sorted(self._collect_reads(wildcards, wildcards.library_name, ""))
+                    )
+                },
+            }
+            reads_right = list(
+                sorted(self._collect_reads(wildcards, wildcards.library_name, "right-"))
+            )
+            if reads_right:
+                result["input"]["reads_right"] = reads_right
+            return result
+
+        assert action == "run", "Unsupported actions"
+        return args_function
+
+    def _collect_reads(self, wildcards, library_name, prefix):
+        """Yield the path to reads
+
+        Yields paths to right reads if prefix=='right-'
+        """
+        _ = library_name
+        folder_name = get_ngs_library_folder_name(self.parent.sheets, wildcards.library_name)
+        pattern_set_keys = ("right",) if prefix.startswith("right-") else ("left",)
+        for _, path_infix, filename in self.path_gen.run(folder_name, pattern_set_keys):
+            yield os.path.join(self.base_path_in, path_infix, filename).format(**wildcards)
+
+
+class BbdukStepPart(AdapterTrimmingStepPart):
+    """bbduk adapter trimming"""
+
+    name = "bbduk"
+
+    def get_resource_usage(self, action):
+        """Get Resource Usage
+
+        :param action: Action (i.e., step) in the workflow, example: 'run'.
+        :type action: str
+
+        :return: Returns ResourceUsage for step.
+
+        :raises UnsupportedActionException: if action not in class defined list of valid actions.
+        """
+        if action not in self.actions:
```

Comment:
Simplified action validation:
```
# Validate action
self._validate_action(action)
```

Reply:
Remove `raises` after change.

Reply:
Still needs attention.
if you have `self._validate_action(action)`, you don't need the `if action not in self.actions` block.

Reply:
Solved in #222